### PR TITLE
backupccl: add more fine-grained privilege checks to BACKUP

### LINKED ISSUE #54635: backupccl: remove admin check for certain flavors of backups
With the move to operator roles and overall goal of removing what needs admin, one of the features we haven't touched yes is BACKUP. Users have a need for these, for example, to seed test environments or to take their own backups.

There are already permission checks for BACKUP but they are superseded by the admin check. We want to allow it such that in the future non-admins (specifically Cockroach Cloud users) can take backups.

** Deliverables **
- Admin check remains for same set of storage as the recent IMPORT work (I believe it's implicit creds with cloud storage, http, https, custom endpoints nodelocal)
- Full cluster and tenant backups still require admin
- Check permissions for user (database - select, table - select, UDT and UDS - usage)

** Additional Notes ** 
- There may be some rough edges here that I'm not aware of -- for example if you have select on database does that also give you usage on all the UDS and UDT within it? If not we get some weird behaviors. In those cases I think it's safest to error out and we can smooth out the rough edges later rather than give a user access to data they shouldn't have access to.

### PR DESCRIPTION
**Pull request by adityamaru**
This changes removes the umbrella admin check from the BACKUP
code path and introduces more role based privilege checking.
This is in keeping with the changes which have recently been made
to IMPORT and RESTORE.

BACKUPS now have the following privilege checks:
- Cluster BACKUPs can only be performed by admin users.
- Tenant BACKUP can only be performed by admin users.
- ExternalStorage URIs which require an admin access
can only be used as a BACKUP source if the user is an admin.
- Database and table BACKUP require SELECT privileges on each respectively.
- UDT, UDS require USAGE privileges on each respectively.

Fixes: #54635

Release note: None

### CODE REVIEW

**Comment by adityamaru** [comment]
Note: UDT and UDS lines of the data-driven test are commented out until #54662 is resolved.

File: pkg/ccl/backupccl/testdata/backup-restore/backup-permissions [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+BACKUP TO 'nodelocal://0/test-nonroot-cluster';
+BACKUP DATABASE d TO 'nodelocal://0/test-nonroot-db';
+BACKUP TABLE d.t TO 'nodelocal://0/test-nonroot-table';
+----
+
+# Start a new cluster with the same IO dir.
+new-server name=s2 allow-implicit-access
+----
+
+exec-sql server=s2
```
Comment (pbardea) [inline pkg/ccl/backupccl/testdata/backup-restore/backup-permissions:32]: nit: It looks like `exec-sql` uses the last created server, so we don't need to specify `server=s2` and `server=s3` on all of the following exec-sql commands.

**Review by pbardea — APPROVED** [review]
LGTM, thanks!

**Comment by pbardea** [comment]
It also looks like the UDT/UDS tests can be uncommented now, right?

File: pkg/ccl/backupccl/testdata/backup-restore/backup-permissions (hunk 1 above)
Comment (adityamaru) [reply pkg/ccl/backupccl/testdata/backup-restore/backup-permissions]: done!

**Comment by adityamaru** [comment]
> It also looks like the UDT/UDS tests can be uncommented now, right? (After a rebase.)

yep, was just clarifying whether UDTs get USAGE on creation. Turns out they do, so I had to revoke to test our BACKUP checks, before re-granting. Uncommented now.

File: pkg/ccl/backupccl/backup_planning.go [hunk 2]
```diff
@@ -433,6 +433,64 @@ func getBackupStatement(stmt tree.Statement) *annotatedBackupStatement {
...
+				return err
+			}
+		case catalog.TypeDescriptor, catalog.SchemaDescriptor:
+			if err := p.CheckPrivilege(ctx, desc, privilege.USAGE); err != nil {
+				return err
+			}
+		}
+	}
+	knobs := p.ExecCfg().BackupRestoreTestingKnobs
+	if knobs == nil || !knobs.AllowImplicitAccess {
```
Comment (dt) [inline pkg/ccl/backupccl/backup_planning.go:475]: nit: I'd do `if knobs != nil && knobs... { return nil } so the common case (no knobs) isn't in a conditional.

**Review by dt — APPROVED** [review]

File: pkg/ccl/backupccl/backup_planning.go (hunk 2 above)
Comment (adityamaru) [reply pkg/ccl/backupccl/backup_planning.go]: done.

**Comment by adityamaru** [comment]
@mwang1026 and I decided that it is okay to wait till the first dot release, removing backport tags.

TFTRs!

bors r=dt,pbardea

**Pull request merged**